### PR TITLE
sink(ticdc): add topic manager

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/4423

### What is changed and how it works?

- add topic manager. It is responsible for creating and tracking information about topics.
- Add GetTopic and GetPartition methods to event router.

### Check List 

Tests 

 - [x] Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

None

Related changes

None

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by hi-rustin** [comment]
- [ ] err
- [ ] log
- [ ] close

**Review by hi-rustin — COMMENTED** [review]
Self check

File: cdc/sink/dispatcher/event_router.go [hunk 1]
```diff
@@ -107,13 +107,23 @@ func NewEventRouter(cfg *config.ReplicaConfig, defaultTopic string) (*EventRoute
...
 }
 
-// DispatchRowChangedEvent returns the target topic and partition.
-func (s *EventRouter) DispatchRowChangedEvent(row *model.RowChangedEvent, partitionNum int32) (string, int32) {
-	topicDispatcher, partitionDispatcher := s.matchDispatcher(row.Table.Schema, row.Table.Table)
-	topic := topicDispatcher.DispatchRowChangedEvent(row)
-	partition := partitionDispatcher.DispatchRowChangedEvent(row, partitionNum)
+// GetTopic returns the target topic.
+func (s *EventRouter) GetTopic(row *model.RowChangedEvent) string {
+	topicDispatcher, _ := s.matchDispatcher(row.Table.Schema, row.Table.Table)
```
Comment (zhaoxinyu) [inline cdc/sink/dispatcher/event_router.go:112]: How about returning a rule and accessing topicDispatcher via `rule.Dispatcher`?

File: cdc/sink/dispatcher/event_router.go [hunk 2]
```diff
@@ -107,13 +107,23 @@ func NewEventRouter(cfg *config.ReplicaConfig, defaultTopic string) (*EventRoute
...
+	return topicDispatcher.DispatchRowChangedEvent(row)
+}
 
-	return topic, partition
+// GetPartition returns the target partition.
+func (s *EventRouter) GetPartition(
+	row *model.RowChangedEvent,
+	partitionNum int32,
+) int32 {
+	_, partitionDispatcher := s.matchDispatcher(row.Table.Schema,
```
Comment (zhaoxinyu) [inline cdc/sink/dispatcher/event_router.go:121]: Ditto

File: cdc/sink/manager/kafka/manager.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		if err != nil {
+			return err
+		}
+
+		for _, topic := range topics {
+			partitions, err := m.client.Partitions(topic)
+			if err != nil {
+				return err
+			}
+			m.topics.Store(topic, int32(len(partitions)))
```
Comment (zhaoxinyu) [inline cdc/sink/manager/kafka/manager.go:84]: Is it necessary to add log messages indicating that the number of partition has been changed?

File: cdc/sink/manager/kafka/manager_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	cfg := &kafkaconfig.AutoCreateTopicConfig{
+		AutoCreate:        true,
+		PartitionNum:      2,
+		ReplicationFactor: 1,
+	}
+
+	manager := NewTopicManager(client, adminClient, cfg)
+	partitionsNum, err := manager.Partitions(kafkamock.DefaultMockTopicName)
+	require.Nil(t, err)
+	require.Equal(t, int32(3), partitionsNum)
```
Comment (zhaoxinyu) [inline cdc/sink/manager/kafka/manager_test.go:42]: Does it mean that `parititionsNum` is 3 in the beginning, but after 1min, it will be changed to 2?

Comment (hi-rustin) [reply cdc/sink/manager/kafka/manager_test.go:42]: No, this is just a test of its initial value.

**Review by zhaoxinyu — APPROVED** [review]

File: cdc/sink/manager/kafka/manager.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+// tryUpdatePartitionsAndLogging try to update the partitions of the topic.
+func (m *TopicManager) tryUpdatePartitionsAndLogging(topic string, partitions int32) {
+	oldPartitions, ok := m.topics.Load(topic)
+	if ok {
+		if oldPartitions.(int32) != partitions {
+			m.topics.Store(topic, partitions)
+			log.Info(
+				"update topic partition number",
+				zap.String("topic", topic),
+				zap.Int32("oldPartitionNumber", oldPartitions.(int32)),
```
Comment (hi-rustin) [inline cdc/sink/manager/kafka/manager.go:101]:
```suggestion
				zap.Int32("old-partition-num", oldPartitions.(int32)),
```

File: cdc/sink/manager/kafka/manager.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+				zap.Int32("newPartitionNumber", partitions),
```
Comment (hi-rustin) [inline cdc/sink/manager/kafka/manager.go:102]:
```suggestion
				zap.Int32("new-partition-numr", partitions),
```

File: cdc/sink/manager/kafka/manager.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+			)
+		}
+	} else {
+		m.topics.Store(topic, partitions)
+		log.Info(
+			"store topic partition number",
+			zap.String("topic", topic),
+			zap.Int32("partitionNumber", partitions),
```
Comment (hi-rustin) [inline cdc/sink/manager/kafka/manager.go:110]:
```suggestion
			zap.Int32("partition-num", partitions),
```

File: cdc/sink/manager/kafka/manager.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		return nil
+	}
+
+	if !m.cfg.AutoCreate {
+		return cerror.ErrKafkaInvalidConfig.GenWithStack(
+			fmt.Sprintf("`auto-create-topic` is false, "+
+				"and %s not found", topicName))
+	}
+
+	err = m.admin.CreateTopic(topicName, &sarama.TopicDetail{
```
Comment (overvenus) [inline cdc/sink/manager/kafka/manager.go:140]: Could you add some logs to track how long these I/O operations takes? Also for list and delete topic.

Comment (hi-rustin) [reply cdc/sink/manager/kafka/manager.go:151]: Done.

**Comment by hi-rustin** [comment]
Thanks for your review!  💚 💙 💜 💛 ❤️

File: cdc/sink/manager/kafka/manager.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+// CreateTopic creates a topic with the given name.
+func (m *TopicManager) CreateTopic(topicName string) error {
+	start := time.Now()
+	topics, err := m.admin.ListTopics()
+	if err != nil {
+		log.Error(
+			"Kafka admin client list topics failed",
+			zap.Error(err),
+			zap.Duration("cost", time.Since(start)),
```
Comment (overvenus) [inline cdc/sink/manager/kafka/manager.go:123]:
```suggestion
			zap.Duration("duration", time.Since(start)),
```

**Review by overvenus — APPROVED** [review]

**Pull request merged**